File: android-app/app/src/test/java/com/claudecode/remote/update/GitHubDownloadAcceleratorTest.kt
package com.claudecode.remote.update

import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import java.net.URI
import java.util.concurrent.TimeUnit

class GitHubDownloadAcceleratorTest {
    @Test
    fun buildGitHubDownloadCandidatesAddsHttpsMirrorsForGitHubReleaseAssets() {
        val originalUrl = "https://github.com/714307168/AgentFlow/releases/download/v1.2.50/AgentFlow-android-1.2.50-134-release.apk"
        val candidates = buildGitHubDownloadCandidates(
            originalUrl,
            listOf(
                "https://fast.example/{url}",
                "http://unsafe.example/{url}",
                "https://encoded.example/download?target={encodedUrl}"
            )
        )

        assertEquals(
            listOf(
                originalUrl,
                "https://fast.example/https://github.com/714307168/AgentFlow/releases/download/v1.2.50/AgentFlow-android-1.2.50-134-release.apk",
                "https://encoded.example/download?target=https%3A%2F%2Fgithub.com%2F714307168%2FAgentFlow%2Freleases%2Fdownload%2Fv1.2.50%2FAgentFlow-android-1.2.50-134-release.apk"
            ),
            candidates.map { it.url }
        )
    }

    @Test
    fun buildGitHubDownloadCandidatesDoesNotMirrorRelayDownloads() {
        val originalUrl = "https://relay.example/downloads/AgentFlow.apk"

        assertEquals(
            listOf(originalUrl),
            buildGitHubDownloadCandidates(
                originalUrl,
                listOf("https://fast.example/{url}")
            ).map { it.url }
        )
    }

    @Test
    fun isGitHubReleaseDownloadUrlOnlyAcceptsGitHubReleaseAssets() {
        assertTrue(isGitHubReleaseDownloadUrl("https://github.com/714307168/AgentFlow/releases/download/v1.2.50/app.apk"))
        assertFalse(isGitHubReleaseDownloadUrl("https://github.com/714307168/AgentFlow/archive/refs/tags/v1.2.50.zip"))
        assertFalse(isGitHubReleaseDownloadUrl("https://example.com/714307168/AgentFlow/releases/download/v1.2.50/app.apk"))
    }

    @Test
    fun selectFastestChoosesFastestHealthyMirror() = kotlinx.coroutines.runBlocking {
        val fastServer = MockWebServer()
        val slowServer = MockWebServer()
        val originalServer = MockWebServer()
        try {
            fastServer.enqueue(MockResponse().setResponseCode(200))
            slowServer.enqueue(MockResponse().setHeadersDelay(150, TimeUnit.MILLISECONDS).setResponseCode(200))
            originalServer.enqueue(MockResponse().setHeadersDelay(80, TimeUnit.MILLISECONDS).setResponseCode(200))
            fastServer.start()
            slowServer.start()
            originalServer.start()

            val originalUrl = "https://github.com/714307168/AgentFlow/releases/download/v1.2.50/app.apk"
            val accelerator = GitHubDownloadAccelerator(
                client = OkHttpClient.Builder()
                    .connectTimeout(1, TimeUnit.SECONDS)
                    .readTimeout(1, TimeUnit.SECONDS)
                    .callTimeout(1, TimeUnit.SECONDS)
                    .build(),
                mirrorTemplates = listOf(
                    "https://slow.example/{encodedUrl}",
                    "https://fast.example/{encodedUrl}"
                ),
                probeRequestFactory = { candidate ->
                    val probeUrl = when (URI(candidate.url).host) {
                        "slow.example" -> slowServer.url("/slow").toString()
                        "fast.example" -> fastServer.url("/fast").toString()
                        else -> originalServer.url("/original").toString()
                    }
                    Request.Builder().url(probeUrl).head().build()
                }
            )

            val selected = accelerator.selectFastest(originalUrl)

            assertEquals("fast.example", URI(selected.url).host)
        } finally {
            fastServer.shutdown()
            slowServer.shutdown()
            originalServer.shutdown()
        }
    }

    @Test
    fun selectFastestFallsBackToGitHubWhenEveryProbeFails() = kotlinx.coroutines.runBlocking {
        val originalUrl = "https://github.com/714307168/AgentFlow/releases/download/v1.2.50/app.apk"
        val accelerator = GitHubDownloadAccelerator(
            client = OkHttpClient(),
            mirrorTemplates = listOf("https://broken.example/{url}"),
            probeRequestFactory = { candidate ->
                Request.Builder().url(candidate.url).head().build()
            }
        )

        val selected = accelerator.selectFastest(originalUrl)

        assertEquals(originalUrl, selected.url)
        assertTrue(selected.original)
    }
}

File: local-agent/test/github-download-accelerator.test.js
const test = require("node:test");
const assert = require("node:assert/strict");

const {
  buildGitHubDownloadCandidates,
  getConfiguredGitHubDownloadMirrorTemplates,
  isGitHubReleaseDownloadUrl,
  selectFastestGitHubDownloadUrl,
} = require("../dist/src/github-download-accelerator.js");

test("buildGitHubDownloadCandidates adds HTTPS mirrors only for GitHub release assets", () => {
  const originalUrl = "https://github.com/714307168/AgentFlow/releases/download/v1.1.207/AgentFlow-1.1.207-x64-setup.exe";
  const candidates = buildGitHubDownloadCandidates(originalUrl, [
    "https://fast.example/{url}",
    "http://unsafe.example/{url}",
    "https://encoded.example/download?target={encodedUrl}",
  ]);

  assert.deepEqual(candidates.map((candidate) => candidate.url), [
    originalUrl,
    "https://fast.example/https://github.com/714307168/AgentFlow/releases/download/v1.1.207/AgentFlow-1.1.207-x64-setup.exe",
    "https://encoded.example/download?target=https%3A%2F%2Fgithub.com%2F714307168%2FAgentFlow%2Freleases%2Fdownload%2Fv1.1.207%2FAgentFlow-1.1.207-x64-setup.exe",
  ]);

  assert.deepEqual(buildGitHubDownloadCandidates("https://relay.example/download/app.exe", [
    "https://fast.example/{url}",
  ]).map((candidate) => candidate.url), ["https://relay.example/download/app.exe"]);
});

test("isGitHubReleaseDownloadUrl only accepts GitHub release download URLs", () => {
  assert.equal(isGitHubReleaseDownloadUrl("https://github.com/714307168/AgentFlow/releases/download/v1.1.207/app.exe"), true);
  assert.equal(isGitHubReleaseDownloadUrl("https://github.com/714307168/AgentFlow/archive/refs/tags/v1.1.207.zip"), false);
  assert.equal(isGitHubReleaseDownloadUrl("https://example.com/714307168/AgentFlow/releases/download/v1.1.207/app.exe"), false);
});

test("getConfiguredGitHubDownloadMirrorTemplates supports custom mirrors and disable switch", () => {
  const previousMirrors = process.env.AGENTFLOW_GITHUB_DOWNLOAD_MIRRORS;
  const previousDisabled = process.env.AGENTFLOW_DISABLE_GITHUB_DOWNLOAD_ACCELERATORS;
  try {
    process.env.AGENTFLOW_GITHUB_DOWNLOAD_MIRRORS = "https://custom-one.example/{url}, https://custom-two.example/{encodedUrl}";
    delete process.env.AGENTFLOW_DISABLE_GITHUB_DOWNLOAD_ACCELERATORS;
    const enabled = getConfiguredGitHubDownloadMirrorTemplates();
    assert.equal(enabled[0], "https://custom-one.example/{url}");
    assert.equal(enabled[1], "https://custom-two.example/{encodedUrl}");
    assert.equal(enabled.some((template) => template.includes("gh.llkk.cc")), true);

    process.env.AGENTFLOW_DISABLE_GITHUB_DOWNLOAD_ACCELERATORS = "1";
    assert.deepEqual(getConfiguredGitHubDownloadMirrorTemplates(), [
      "https://custom-one.example/{url}",
      "https://custom-two.example/{encodedUrl}",
    ]);
  } finally {
    restoreEnv("AGENTFLOW_GITHUB_DOWNLOAD_MIRRORS", previousMirrors);
    restoreEnv("AGENTFLOW_DISABLE_GITHUB_DOWNLOAD_ACCELERATORS", previousDisabled);
  }
});

test("selectFastestGitHubDownloadUrl chooses the fastest healthy candidate", async () => {
  const originalUrl = "https://github.com/714307168/AgentFlow/releases/download/v1.1.207/AgentFlow-1.1.207-x64-setup.exe";
  const calls = [];
  const selected = await selectFastestGitHubDownloadUrl(originalUrl, {
    timeoutMs: 1000,
    mirrorTemplates: [
      "https://slow.example/{url}",
      "https://fast.example/{url}",
    ],
    fetchImpl: async (url) => {
      calls.push(String(url));
      if (String(url).includes("slow.example")) {
        await delay(40);
        return new Response(null, { status: 200 });
      }
      if (String(url).includes("fast.example")) {
        await delay(5);
        return new Response(null, { status: 200 });
      }
      await delay(20);
      return new Response(null, { status: 200 });
    },
  });

  assert.equal(selected.url.startsWith("https://fast.example/"), true);
  assert.equal(calls.length, 3);
});

test("selectFastestGitHubDownloadUrl falls back to GitHub when every probe fails", async () => {
  const originalUrl = "https://github.com/714307168/AgentFlow/releases/download/v1.1.207/AgentFlow-1.1.207-x64-setup.exe";
  const selected = await selectFastestGitHubDownloadUrl(originalUrl, {
    timeoutMs: 100,
    mirrorTemplates: ["https://broken.example/{url}"],
    fetchImpl: async () => new Response(null, { status: 503 }),
  });

  assert.equal(selected.url, originalUrl);
  assert.equal(selected.original, true);
});

function delay(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function restoreEnv(name, value) {
  if (value === undefined) {
    delete process.env[name];
  } else {
    process.env[name] = value;
  }
}
